protocols/bftcosi/bftcosi.go

```diff
@@ -187,291 +200,300 @@ func (bft *ProtocolBFTCoSi) Dispatch() error {
 	}
 }
 
-// startAnnouncementPrepare create its announcement for the prepare round and
-// sends it down the tree.
-func (bft *ProtocolBFTCoSi) startAnnouncementPrepare() error {
-	ann := bft.prepare.CreateAnnouncement()
-	a := &Announce{
-		TYPE:         RoundPrepare,
-		Announcement: ann,
+// Signature will generate the final signature, the output of the BFTCoSi
+// protocol.
+// The signature contains the commit round signature, with the message.
+// If the prepare phase failed, the signature will be nil and the Exceptions
+// will contain the exception from the prepare phase. It can be useful to see
+// which cosigners refused to sign (each exceptions contains the index of a
+// refusing-to-sign signer).
+// Expect this function to have an undefined behavior when called from a
+// non-root Node.
+func (bft *ProtocolBFTCoSi) Signature() *BFTSignature {
+	bftSig := &BFTSignature{
+		Sig:        bft.commit.Signature(),
+		Msg:        bft.Msg,
+		Exceptions: nil,
 	}
-	dbg.Lvl4("BFTCoSi Start Announcement (PREPARE)")
-	return bft.sendAnnouncement(a)
+	if bft.signRefusal {
+		bftSig.Sig = nil
+		bftSig.Exceptions = bft.tempExceptions
+	}
+	return bftSig
+}
+
+// RegisterOnDone registers a callback to call when the bftcosi protocols has
+// really finished
+func (bft *ProtocolBFTCoSi) RegisterOnDone(fn func()) {
+	bft.onDoneCallback = fn
 }
 
-// startAnnouncementCommit create the announcement for the commit phase and
+// RegisterOnSignatureDone register a callback to call when the bftcosi
+// protocol reached a signature on the block
+func (bft *ProtocolBFTCoSi) RegisterOnSignatureDone(fn func(*BFTSignature)) {
```

Comment:
I think this must be removed - I don't know how it re-appeared here, pretty sure I took it out at _some_ time.